File: NetworkSensorAppJava/app/src/main/java/com/networksensor/app/repository/NetworkRepository.java
package com.networksensor.app.repository;

import android.annotation.SuppressLint;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkCapabilities;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.telephony.TelephonyManager;
import android.util.Log;

import com.networksensor.app.model.NetworkInfo;
import com.networksensor.app.model.NetworkInfo.ConnectionType;

import java.io.IOException;
import java.net.InetAddress;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class NetworkRepository {
    private static final String TAG = "NetworkRepository";
    private final Context context;
    private final OkHttpClient okHttpClient;
    private final ConnectivityManager connectivityManager;
    private final WifiManager wifiManager;
    private final TelephonyManager telephonyManager;
    private final ExecutorService executorService;

    public NetworkRepository(Context context) {
        this.context = context;
        this.okHttpClient = new OkHttpClient.Builder()
                .connectTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
                .readTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
                .writeTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
                .build();
        
        this.connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        this.wifiManager = (WifiManager) context.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        this.telephonyManager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
        this.executorService = Executors.newFixedThreadPool(4);
    }

    public Observable<NetworkInfo> getCurrentNetworkInfo() {
        return Observable.fromCallable(this::getCurrentNetworkInfoSync)
                .subscribeOn(Schedulers.io());
    }

    @SuppressLint("MissingPermission")
    private NetworkInfo getCurrentNetworkInfoSync() {
        try {
            android.net.Network activeNetwork = connectivityManager.getActiveNetwork();
            NetworkCapabilities networkCapabilities = connectivityManager.getNetworkCapabilities(activeNetwork);
            
            ConnectionType connectionType = determineConnectionType(networkCapabilities);
            String networkName = getNetworkName(connectionType);
            int signalStrength = getSignalStrength(connectionType);
            
            // Perform quick speed test
            SpeedTestResult speedTest = performQuickSpeedTest();
            
            return new NetworkInfo(
                connectionType,
                networkName,
                signalStrength,
                speedTest.downloadSpeed,
                speedTest.uploadSpeed,
                speedTest.latency
            );
        } catch (Exception e) {
            Log.e(TAG, "Error getting network info", e);
            return new NetworkInfo(ConnectionType.UNKNOWN, "Unknown", 0, 0.0, 0.0, 0L);
        }
    }

    private ConnectionType determineConnectionType(NetworkCapabilities networkCapabilities) {
        if (networkCapabilities == null) {
            return ConnectionType.UNKNOWN;
        }

        if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
            return ConnectionType.WIFI;
        } else if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
            return getCellularConnectionType();
        } else if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
            return ConnectionType.ETHERNET;
        }
        
        return ConnectionType.UNKNOWN;
    }

    @SuppressLint("MissingPermission")
    private ConnectionType getCellularConnectionType() {
        try {
            int networkType = telephonyManager.getDataNetworkType();
            switch (networkType) {
                case TelephonyManager.NETWORK_TYPE_NR:
                    return ConnectionType.MOBILE_5G;
                case TelephonyManager.NETWORK_TYPE_LTE:
                    return ConnectionType.MOBILE_LTE;
                case TelephonyManager.NETWORK_TYPE_HSPAP:
                case TelephonyManager.NETWORK_TYPE_HSPA:
                case TelephonyManager.NETWORK_TYPE_HSUPA:
                case TelephonyManager.NETWORK_TYPE_HSDPA:
                case TelephonyManager.NETWORK_TYPE_UMTS:
                    return ConnectionType.MOBILE_3G;
                case TelephonyManager.NETWORK_TYPE_EDGE:
                case TelephonyManager.NETWORK_TYPE_GPRS:
                    return ConnectionType.MOBILE_2G;
                default:
                    return ConnectionType.MOBILE_4G;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error determining cellular connection type", e);
            return ConnectionType.MOBILE_4G;
        }
    }

    @SuppressLint("MissingPermission")
    private String getNetworkName(ConnectionType connectionType) {
        try {
            if (connectionType == ConnectionType.WIFI) {
                WifiInfo wifiInfo = wifiManager.getConnectionInfo();
                String ssid = wifiInfo.getSSID();
                return ssid != null ? ssid.replace("\"", "") : "Unknown WiFi";
            } else {
                String operatorName = telephonyManager.getNetworkOperatorName();
                return operatorName != null && !operatorName.isEmpty() ? operatorName : "Unknown Carrier";
            }
        } catch (Exception e) {
            Log.e(TAG, "Error getting network name", e);
            return "Unknown";
        }
    }

    @SuppressLint("MissingPermission")
    private int getSignalStrength(ConnectionType connectionType) {
        try {
            if (connectionType == ConnectionType.WIFI) {
                WifiInfo wifiInfo = wifiManager.getConnectionInfo();
                return WifiManager.calculateSignalLevel(wifiInfo.getRssi(), 5);
            } else {
                // For cellular, we would need more complex implementation
                // For now, return a default value
                return 3;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error getting signal strength", e);
            return 0;
        }
    }

    public Observable<SpeedTestResult> performSpeedTest() {
        return Observable.fromCallable(() -> {
            String[] testUrls = {
                "https://httpbin.org/bytes/1000000", // 1MB
                "https://httpbin.org/bytes/5000000", // 5MB
                "https://httpbin.org/bytes/10000000" // 10MB
            };

            double totalDownloadSpeed = 0.0;
            long totalLatency = 0L;
            int successfulTests = 0;

            for (String url : testUrls) {
                try {
                    SpeedTestResult result = performSingleSpeedTest(url);
                    totalDownloadSpeed += result.downloadSpeed;
                    totalLatency += result.latency;
                    successfulTests++;
                } catch (Exception e) {
                    Log.w(TAG, "Speed test failed for URL: " + url, e);
                }
            }

            if (successfulTests > 0) {
                return new SpeedTestResult(
                    totalDownloadSpeed / successfulTests,
                    0.0, // Upload speed would require different endpoint
                    totalLatency / successfulTests,
                    0L, // Jitter calculation would require multiple tests
                    0.0, // Packet loss calculation
                    System.currentTimeMillis()
                );
            } else {
                throw new IOException("All speed tests failed");
            }
        }).subscribeOn(Schedulers.io());
    }

    private SpeedTestResult performQuickSpeedTest() {
        try {
            return performSingleSpeedTest("https://httpbin.org/bytes/1000000");
        } catch (Exception e) {
            Log.e(TAG, "Quick speed test failed", e);
            return new SpeedTestResult(0.0, 0.0, 0L, 0L, 0.0, System.currentTimeMillis());
        }
    }

    private SpeedTestResult performSingleSpeedTest(String url) throws IOException {
        long latency = measureLatency("8.8.8.8");
        
        Request request = new Request.Builder().url(url).build();
        
        long startTime = System.currentTimeMillis();
        try (Response response = okHttpClient.newCall(request).execute()) {
            if (response.isSuccessful() && response.body() != null) {
                byte[] bytes = response.body().bytes();
                long downloadTime = System.currentTimeMillis() - startTime;
                
                // Calculate download speed in Mbps
                double fileSizeBytes = bytes.length;
                double downloadSpeedMbps = (fileSizeBytes * 8.0) / (downloadTime * 1000.0);
                
                return new SpeedTestResult(
                    downloadSpeedMbps,
                    0.0, // Upload speed
                    latency,
                    0L, // Jitter
                    0.0, // Packet loss
                    downloadTime
                );
            } else {
                throw new IOException("Response not successful: " + response.code());
            }
        }
    }

    private long measureLatency(String host) {
        try {
            long startTime = System.currentTimeMillis();
            InetAddress.getByName(host).isReachable(5000);
            return System.currentTimeMillis() - startTime;
        } catch (Exception e) {
            Log.w(TAG, "Latency measurement failed for host: " + host, e);
            return 0L;
        }
    }

    public void cleanup() {
        if (executorService != null && !executorService.isShutdown()) {
            executorService.shutdown();
        }
    }

    // Inner class for speed test results
    public static class SpeedTestResult {
        public final double downloadSpeed;
        public final double uploadSpeed;
        public final long latency;
        public final long jitter;
        public final double packetLoss;
        public final long testDuration;

        public SpeedTestResult(double downloadSpeed, double uploadSpeed, long latency,
                              long jitter, double packetLoss, long testDuration) {
            this.downloadSpeed = downloadSpeed;
            this.uploadSpeed = uploadSpeed;
            this.latency = latency;
            this.jitter = jitter;
            this.packetLoss = packetLoss;
            this.testDuration = testDuration;
        }
    }
}

File: NetworkSensorApp/app/src/main/java/com/networksensor/app/data/repository/NetworkRepository.kt
package com.networksensor.app.data.repository

import android.annotation.SuppressLint
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.net.wifi.WifiManager
import android.telephony.TelephonyManager
import android.telephony.SignalStrength
import com.networksensor.app.data.model.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.net.InetAddress
import java.net.URL
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.system.measureTimeMillis

@Singleton
class NetworkRepository @Inject constructor(
    private val context: Context,
    private val okHttpClient: OkHttpClient
) {
    
    private val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    private val wifiManager = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
    private val telephonyManager = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
    
    fun getCurrentNetworkInfo(): Flow<NetworkInfo> = flow {
        val networkInfo = getCurrentNetworkInfoSync()
        emit(networkInfo)
    }.flowOn(Dispatchers.IO)
    
    @SuppressLint("MissingPermission")
    private suspend fun getCurrentNetworkInfoSync(): NetworkInfo = withContext(Dispatchers.IO) {
        val activeNetwork = connectivityManager.activeNetwork
        val networkCapabilities = connectivityManager.getNetworkCapabilities(activeNetwork)
        
        val connectionType = when {
            networkCapabilities?.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) == true -> ConnectionType.WIFI
            networkCapabilities?.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) == true -> {
                when (telephonyManager.dataNetworkType) {
                    TelephonyManager.NETWORK_TYPE_NR -> ConnectionType.MOBILE_5G
                    TelephonyManager.NETWORK_TYPE_LTE -> ConnectionType.MOBILE_LTE
                    TelephonyManager.NETWORK_TYPE_HSPAP,
                    TelephonyManager.NETWORK_TYPE_HSPA,
                    TelephonyManager.NETWORK_TYPE_HSUPA,
                    TelephonyManager.NETWORK_TYPE_HSDPA,
                    TelephonyManager.NETWORK_TYPE_UMTS -> ConnectionType.MOBILE_3G
                    TelephonyManager.NETWORK_TYPE_EDGE,
                    TelephonyManager.NETWORK_TYPE_GPRS -> ConnectionType.MOBILE_2G
                    else -> ConnectionType.MOBILE_4G
                }
            }
            networkCapabilities?.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) == true -> ConnectionType.ETHERNET
            else -> ConnectionType.UNKNOWN
        }
        
        val networkName = when (connectionType) {
            ConnectionType.WIFI -> getWifiNetworkName()
            else -> telephonyManager.networkOperatorName
        }
        
        val signalStrength = getSignalStrength(connectionType)
        val speedTest = performQuickSpeedTest()
        
        NetworkInfo(
            connectionType = connectionType,
            networkName = networkName,
            signalStrength = signalStrength,
            downloadSpeed = speedTest.downloadSpeed,
            uploadSpeed = speedTest.uploadSpeed,
            latency = speedTest.latency
        )
    }
    
    @SuppressLint("MissingPermission")
    private fun getWifiNetworkName(): String? {
        return try {
            val wifiInfo = wifiManager.connectionInfo
            wifiInfo.ssid?.replace("\"", "")
        } catch (e: Exception) {
            null
        }
    }
    
    @SuppressLint("MissingPermission")
    private fun getSignalStrength(connectionType: ConnectionType): Int {
        return when (connectionType) {
            ConnectionType.WIFI -> {
                try {
                    val wifiInfo = wifiManager.connectionInfo
                    WifiManager.calculateSignalLevel(wifiInfo.rssi, 5)
                } catch (e: Exception) {
                    0
                }
            }
            else -> {
                try {
                    // This would require more complex implementation for cellular signal strength
                    3 // Default value
                } catch (e: Exception) {
                    0
                }
            }
        }
    }
    
    suspend fun performSpeedTest(): Flow<SpeedTestResult> = flow {
        val testUrls = listOf(
            "https://speed.cloudflare.com/__down?bytes=25000000", // 25MB
            "https://httpbin.org/bytes/10000000", // 10MB
            "https://www.google.com/images/branding/googlelogo/1x/googlelogo_color_272x92dp.png"
        )
        
        var totalDownloadSpeed = 0.0
        var totalLatency = 0L
        var successfulTests = 0
        
        for (url in testUrls) {
            try {
                val result = performSingleSpeedTest(url)
                totalDownloadSpeed += result.downloadSpeed
                totalLatency += result.latency
                successfulTests++
                
                emit(SpeedTestResult(
                    downloadSpeed = totalDownloadSpeed / successfulTests,
                    uploadSpeed = result.uploadSpeed,
                    latency = totalLatency / successfulTests,
                    jitter = result.jitter,
                    packetLoss = result.packetLoss,
                    testDuration = result.testDuration
                ))
            } catch (e: Exception) {
                // Continue with next test
            }
        }
    }.flowOn(Dispatchers.IO)
    
    private suspend fun performQuickSpeedTest(): SpeedTestResult = withContext(Dispatchers.IO) {
        try {
            performSingleSpeedTest("https://httpbin.org/bytes/1000000") // 1MB for quick test
        } catch (e: Exception) {
            SpeedTestResult(0.0, 0.0, 0L, 0L, 0.0, 0L)
        }
    }
    
    private suspend fun performSingleSpeedTest(url: String): SpeedTestResult = withContext(Dispatchers.IO) {
        val request = Request.Builder().url(url).build()
        
        val latency = measureLatency("8.8.8.8")
        
        val downloadTime = measureTimeMillis {
            try {
                okHttpClient.newCall(request).execute().use { response ->
                    if (response.isSuccessful) {
                        response.body?.bytes()
                    }
                }
            } catch (e: IOException) {
                throw e
            }
        }
        
        // Estimate file size (this would be more accurate with actual file sizes)
        val fileSizeBytes = 1000000 // 1MB
        val downloadSpeedMbps = (fileSizeBytes * 8.0) / (downloadTime * 1000.0) // Convert to Mbps
        
        SpeedTestResult(
            downloadSpeed = downloadSpeedMbps,
            uploadSpeed = 0.0, // Upload test would require a different endpoint
            latency = latency,
            jitter = 0L,
            packetLoss = 0.0,
            testDuration = downloadTime
        )
    }
    
    private suspend fun measureLatency(host: String): Long = withContext(Dispatchers.IO) {
        try {
            val startTime = System.currentTimeMillis()
            InetAddress.getByName(host).isReachable(5000)
            System.currentTimeMillis() - startTime
        } catch (e: Exception) {
            0L
        }
    }
    
    @SuppressLint("MissingPermission")
    fun getWifiDetails(): Flow<WifiInfo> = flow {
        try {
            val wifiInfo = wifiManager.connectionInfo
            val dhcpInfo = wifiManager.dhcpInfo
            
            emit(WifiInfo(
                ssid = wifiInfo.ssid?.replace("\"", ""),
                bssid = wifiInfo.bssid,
                frequency = wifiInfo.frequency,
                linkSpeed = wifiInfo.linkSpeed,
                rssi = wifiInfo.rssi,
                networkId = wifiInfo.networkId,
                ipAddress = intToIp(dhcpInfo.ipAddress)
            ))
        } catch (e: Exception) {
            emit(WifiInfo(null, null, 0, 0, 0, 0, null))
        }
    }.flowOn(Dispatchers.IO)
    
    @SuppressLint("MissingPermission")
    fun getCellularDetails(): Flow<CellularInfo> = flow {
        try {
            emit(CellularInfo(
                networkType = getNetworkTypeString(),
                operatorName = telephonyManager.networkOperatorName,
                mcc = telephonyManager.networkOperator?.substring(0, 3),
                mnc = telephonyManager.networkOperator?.substring(3),
                signalStrength = 3, // Would need more complex implementation
                cellId = null, // Would require location permissions and more complex implementation
                lac = null
            ))
        } catch (e: Exception) {
            emit(CellularInfo("Unknown", null, null, null, 0, null, null))
        }
    }.flowOn(Dispatchers.IO)
    
    private fun getNetworkTypeString(): String {
        return when (telephonyManager.dataNetworkType) {
            TelephonyManager.NETWORK_TYPE_NR -> "5G NR"
            TelephonyManager.NETWORK_TYPE_LTE -> "LTE"
            TelephonyManager.NETWORK_TYPE_HSPAP -> "HSPA+"
            TelephonyManager.NETWORK_TYPE_HSPA -> "HSPA"
            TelephonyManager.NETWORK_TYPE_HSUPA -> "HSUPA"
            TelephonyManager.NETWORK_TYPE_HSDPA -> "HSDPA"
            TelephonyManager.NETWORK_TYPE_UMTS -> "UMTS"
            TelephonyManager.NETWORK_TYPE_EDGE -> "EDGE"
            TelephonyManager.NETWORK_TYPE_GPRS -> "GPRS"
            else -> "Unknown"
        }
    }
    
    private fun intToIp(ip: Int): String {
        return "${ip and 0xFF}.${ip shr 8 and 0xFF}.${ip shr 16 and 0xFF}.${ip shr 24 and 0xFF}"
    }
}